src/main/java/org/springframework/integration/kafka/listener/MetadataStoreOffsetManager.java

```diff
@@ -0,0 +1,207 @@
+	}
+
+	public String getConsumerId() {
+		return consumerId;
+	}
+
+	/**
+	 * The identifier of a consumer of Kafka messages. Allows to store separate sets of offsets in the {@link MetadataStore}.
+	 *
+	 * @param consumerId the consumer ID
+	 */
+	public void setConsumerId(String consumerId) {
+		this.consumerId = consumerId;
+	}
+
+	public MetadataStore getMetadataStore() {
+		return metadataStore;
+	}
+
+	/**
+	 * The backing {@link MetadataStore} for storing offsets.
+	 *
+	 * @param metadataStore a fully configured {@link MetadataStore} instance
+	 */
+	public void setMetadataStore(MetadataStore metadataStore) {
+		this.metadataStore = metadataStore;
+	}
+
+	public long getReferenceTimestamp() {
+		return referenceTimestamp;
+	}
+
+	/**
+	 * A timestamp to be used for resetting initial offsets, if they are not available in the {@link MetadataStore}
+	 *
+	 * @param referenceTimestamp
+	 */
+	public void setReferenceTimestamp(long referenceTimestamp) {
+		this.referenceTimestamp = referenceTimestamp;
+	}
+
+	/**
+	 * @see OffsetManager#updateOffset(Partition, long)
+	 */
+	@Override
+	public void updateOffset(Partition partition, long offset) {
+		metadataStore.put(asKey(partition), Long.toString(offset));
+		offsets.put(partition, offset);
+	}
+
+	/**
+	 * @see OffsetManager#getOffset(Partition)
+	 */
+	@Override
+	public long getOffset(Partition partition) {
+		if(!offsets.containsKey(partition)) {
+			initializeOffsets(Collections.singleton(partition));
+		}
+		if (!offsets.containsKey(partition)) {
+			throw new IllegalStateException("Offset " + offsets + " cannot be found");
+		}
+		return offsets.get(partition);
+	}
+
+	@Override
+	public void resetOffsets(Collection<Partition> partitionsToReset) {
+		MutableMultimap<BrokerAddress, Partition> partitionsByLeaderAddress = groupBy(partitionsToReset, getLeaderFunction);
+		partitionsByLeaderAddress.toMap().forEachKeyValue(new RetrieveInitialOffsetsAndUpdateProcedure());
+	}
+
+	private void initializeOffsets(Collection<Partition> partitions) {
+		List<Partition> partitionsRequiringInitialOffsets = new ArrayList<Partition>();
+		for (Partition partition : partitions) {
+			if (!this.offsets.containsKey(partition)) {
+				String storedOffsetValueAsString = this.metadataStore.get(asKey(partition));
+				Long storedOffsetValue = null;
+				if (storedOffsetValueAsString != null) {
+					try {
+						storedOffsetValue = Long.parseLong(storedOffsetValueAsString);
+					}
+					catch (NumberFormatException e) {
+						LOG.warn("Invalid value: " + storedOffsetValue);
+					}
+				}
+				if (storedOffsetValue != null) {
+					offsets.put(partition, storedOffsetValue);
+				}
+				else {
+					partitionsRequiringInitialOffsets.add(partition);
+				}
+			}
+		}
+		if (partitionsRequiringInitialOffsets.size() > 0) {
+			resetOffsets(partitionsRequiringInitialOffsets);
+		}
+	}
+
+	private String asKey(Partition partition) {
+		return partition.getTopic() + " " + partition.getId() + " " + consumerId;
```

Comment:
Whitespace is a safe separator, since it's not a valid character in a topic name. I don't think `:` is either, but I don;t think this particularly matters.